cub/cub/util_device.cuh

```diff
@@ -69,6 +70,161 @@ CUB_NAMESPACE_BEGIN
 
 #ifndef DOXYGEN_SHOULD_SKIP_THIS // Do not document
 
+namespace detail
+{
+
+/**
+ * @brief Helper class template that allows overwriting the `BLOCK_THREAD` and `ITEMS_PER_THREAD`
+ * configurations of a given policy.
+ */
+template <typename PolicyT, int BLOCK_THREADS_, int ITEMS_PER_THREAD_ = PolicyT::ITEMS_PER_THREAD>
+struct policy_wrapper_t : PolicyT
+{
+  static constexpr int ITEMS_PER_THREAD = ITEMS_PER_THREAD_;
+  static constexpr int BLOCK_THREADS    = BLOCK_THREADS_;
+};
+
+/**
+ * @brief Helper struct to wrap all the information needed to implement virtual shared memory that's passed to a kernel.
+ *
+ */
+struct vsmem_t
+{
+  void* gmem_ptr;
+};
+
+// The maximum amount of shared memory available per thread block
+static constexpr std::size_t max_smem_per_block = 48 * 1024;
+
+/**
+ * @brief Class template that helps to prevent exceeding the available shared memory per thread
+ * block.
+ *
+ * @tparam DefaultAgentPolicyT The default tuning policy that is used
+ * @tparam DefaultAgentT The default agent, instantiated with the given default tuning policy
+ * @tparam FallbackAgentPolicyT A fallback tuning policy that uses a smaller tile size and
+ * therefore exhibits lower shared memory requirements than the default. This fallback policy may be used
+ * before defaulting to global memory-backed virtual shared memory.
+ * @tparam FallbackAgentT The fallback agent, instantiated with the given fallback tuning policy
+ */
+template <typename DefaultAgentPolicyT,
+          typename DefaultAgentT,
+          typename FallbackAgentPolicyT = DefaultAgentPolicyT,
+          typename FallbackAgentT       = DefaultAgentT>
+class vsmem_helper_impl
+{
+public:
+  // True if default agent's shared memory requirements exceed the available shared memory per
+  // thread block but would still suffice for fallback agent's shared memory requirements
+  static constexpr bool uses_fallback_policy =
+    (sizeof(typename DefaultAgentT::TempStorage) > max_smem_per_block)
+    && (sizeof(typename FallbackAgentT::TempStorage) <= max_smem_per_block);
+
+public:
+  // The tuning policy and agent that will be used during algorithm invocation
+  using agent_t        = cub::detail::conditional_t<uses_fallback_policy, FallbackAgentT, DefaultAgentT>;
```

Comment:
If I'm reading this correctly, if both agents exceed max_smem, the default agent will be used with vsmem.

I think that's reasonable, but it should be documented which agent will be used in this case. Can you add this to the class docs?

Reply:
Yup, that's right. I've now added a comment on this to the class docs 👍